docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,358 @@
+}
+
+message SetRetainHeightRequest {
+    int64 height = 1;
+}
+
+message SetRetainHeightResponse {}
+
+message GetRetainHeightRequest {}
+
+message GetRetainHeightResponse {
+    // The retain height as set by the data companion.
+    int64 data_companion_retain_height = 1;
+    // The retain height as set by the ABCI application.
+    int64 app_retain_height = 2;
+}
+```
+
+With this API design, it is technically possible for an integrator to attach
+multiple data companions to the node, but only one of their retain heights will
+be applied.
+
+### Access Control
+
+As covered in the [gRPC API section](#grpc-api), it would be preferable to
+implement some form of access control for sensitive, data companion-specific
+APIs. At least **basic HTTP authentication** should be implemented for these
+endpoints, where credentials should be obtained from an `.htpasswd` file, using
+the same format as [Apache `.htpasswd` files][htpasswd], whose location is set
+in the Tendermint configuration file.
+
+`.htpasswd` files are relatively standard and are supported by many web servers.
+The format is relatively straightforward to parse and interpret.
+
+We should strongly consider only supporting the bcrypt encryption option for
+passwords stored in `.htpasswd` files.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
+companion API.
+
+```toml
+# A data companion, if enabled, is intended to offload the storage of certain
+# types of data from the node. Specifically:
+# 1. Block data, including transactions.
+# 2. Block results, including events, transaction results, validator and
+#    consensus parameter updates.
+#
+# A data companion can influence the pruning height of the node, and therefore
+# the data companion gRPC service is considered to be a sensitive endpoint that
+# is password-protected by default.
+[data_companion]
+
+# Is the data companion gRPC API enabled at all? Default: false
+enabled = false
+
+# The maximum number of blocks to allow the data companion's retain height to
+# lag behind the current height. The node will block all other operations until
+# the companion's retain height is less than `max_retain_lag` blocks behind the
+# current height. Default: 100
+max_retain_lag = 100
+
+# Authentication configuration for the data companion.
+[data_companion.authentication]
+
+# The authentication method to use. At present, the only supported method is
+# "basic" (i.e. basic HTTP authentication).
+method = "basic"
+
+# Path to the file containing basic authentication credentials to access the
+# data companion service. If the data companion is enabled and this password
+# file is not supplied, or the file does not exist, or the file is of an invalid
+# format, the node will fail to start.
+#
+# See https://httpd.apache.org/docs/current/programs/htpasswd.html for details
+# on how to create/configure this file.
+password_file = "/path/to/.htpasswd"
```

Comment:
In terms of UX, I'm not sure what is more cumbersome, whether this, or simply having the companion gRPC separate and bound to _localhost_.